### PR TITLE
[misc] benchmark_throughput : Add LoRA

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
Update benchmark_throughput.py to support LoRA benchmarking.

Examples:
 Machine : 1xA100

`python3 benchmarks/benchmark_throughput.py --model  meta-llama/Llama-2-7b-hf --backend vllm   --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts 1000`

Throughput: 11.45 requests/s, 5509.93 total tokens/s, 2693.62 output tokens/s

`python3 benchmarks/benchmark_throughput.py --model  meta-llama/Llama-2-7b-hf --backend vllm   --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts ${num_prompts} --max-loras 1 --max-lora-rank 8  --enable-lora --lora-path "yard1/llama-2-7b-sql-lora-test"`

Throughput: 7.60 requests/s, 3656.98 total tokens/s, 1787.78 output tokens/s

`python3 benchmarks/benchmark_throughput.py --model  meta-llama/Llama-2-7b-hf --backend vllm   --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts ${num_prompts} --max-loras 4 --max-lora-rank 8  --enable-lora --lora-path "yard1/llama-2-7b-sql-lora-test"`

Throughput: 7.61 requests/s, 3664.67 total tokens/s, 1791.53 output tokens/s

`python3 benchmarks/benchmark_throughput.py --model  meta-llama/Llama-2-7b-hf --backend vllm   --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts ${num_prompts} --async-engine`

Throughput: 11.31 requests/s, 5441.21 total tokens/s, 2660.02 output tokens/s

`python3 benchmarks/benchmark_throughput.py --model  meta-llama/Llama-2-7b-hf --backend vllm   --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts ${num_prompts} --async-engine --max-loras 1 --max-lora-rank 8  --enable-lora  --lora-path "yard1/llama-2-7b-sql-lora-test"`

Throughput: 7.59 requests/s, 3654.39 total tokens/s, 1786.51 output tokens/s

`python3 benchmarks/benchmark_throughput.py --model  meta-llama/Llama-2-7b-hf --backend vllm   --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts ${num_prompts} --async-engine --max-loras 4 --max-lora-rank 8  --enable-lora --lora-path "yard1/llama-2-7b-sql-lora-test"`

Throughput: 7.55 requests/s, 3634.68 total tokens/s, 1776.87 output tokens/s

cc @mgoin @robertgshaw2-neuralmagic @jeejeelee

### CODE REVIEW

File: benchmarks/benchmark_throughput.py [hunk 1]
```diff
@@ -60,8 +66,33 @@ def _get_prompt_for_image_model(question: str, *, model: str) -> str:
     raise ValueError(f"Unsupported model {model}")
 
 
+@cache
+def download_lora(lora_path: str) -> str:
+    """
+    Given an huggingface path, downloads the LoRA model and returns the path
+    on disk.
+    """
+    return snapshot_download(lora_path)
```
Comment (jeejeelee) [inline benchmarks/benchmark_throughput.py:75]: We also shoudl support loading from local path .See: [get_adapter_absolute_path](https://github.com/vllm-project/vllm/blob/v0.6.5/vllm/lora/utils.py#L152)

```suggestion
    return get_adapter_absolute_path(lora_path)
```

File: benchmarks/benchmark_throughput.py [hunk 2]
```diff
@@ -422,6 +481,11 @@ def main(args: argparse.Namespace):
                         action='store_true',
                         default=False,
                         help="Disable decoupled async engine frontend.")
+    # LoRA
+    parser.add_argument("--lora-path",
+                        type=str,
+                        default='yard1/llama-2-7b-sql-lora-test')
```
Comment (jeejeelee) [inline benchmarks/benchmark_throughput.py:487]: Not recommended to use `default` , as LoRA is strongly coupled bound to the model. We also need to add  `help`  to describe this arg

File: benchmarks/benchmark_throughput.py [hunk 3]
```diff
@@ -102,9 +135,14 @@ def sample_requests(tokenizer: PreTrainedTokenizerBase,
                 continue
             prompt = _get_prompt_for_image_model(question=prompt, model=model)
 
+        lora_request: Optional[LoRARequest] = None
+        request_tokenizer = tokenizer
+        if args.enable_lora:
+            lora_request, request_tokenizer = get_random_lora_request(args)
+
```
Comment (jeejeelee) [inline benchmarks/benchmark_throughput.py:142]: There are too many LoRA weights that don't include tokenizer files, which leads to `request_tokenizer` being None. We need to consider this

File: benchmarks/benchmark_throughput.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply benchmarks/benchmark_throughput.py]: Updated to use `get_adapter_absolute_path` :+1:

File: benchmarks/benchmark_throughput.py (hunk 3 above)
Comment (varun-sundar-rabindranath) [reply benchmarks/benchmark_throughput.py:141]: Nice catch ! Added a fallback to use the base tokenizer.

File: benchmarks/benchmark_throughput.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply benchmarks/benchmark_throughput.py]: You are correct. I have removed the default :+1:  Thanks !

**Review by jeejeelee — APPROVED** [review]
overall lgtm, thanks

**Pull request merged**